### ISSUE TITLE
feat(inference): implement sparse-checkpoint replay for the cross-turn GDN cache (ReplayFromCheckpoint is a stub)

### ISSUE THREAD

**Issue by ohdearquant** [ohdearquant/lattice#590]
## Context

The cross-turn KV/GDN prefix-reuse planner (`crates/inference/src/kv_cache/cross_turn.rs`) defines three reuse modes in `PrefixReuseMode` (lines 77-87): `FullRefill`, `ExactAppend`, and `ReplayFromCheckpoint { checkpoint_len }`. Only the first two are reachable at runtime today.

`plan_prefix_reuse` (lines 119-179) already contains real logic for `ReplayFromCheckpoint` — lines 161-176 reuse state up to the highest exact checkpoint boundary at or below the shared token prefix — but the Metal runtime never gives it anything to find. `plan_cross_turn_reuse` in `crates/inference/src/forward/metal_qwen35.rs` (lines 16134-16144) always calls `plan_prefix_reuse(entry, metadata, prompt_ids, &[])` with an empty checkpoint slice, and its own doc comment explains why: "v1 never owns sparse GDN checkpoints, so it always plans with an empty checkpoint set — `PrefixReuseMode::ReplayFromCheckpoint` is therefore never produced by this call."

Even if a checkpoint were supplied, execution isn't implemented either: the `ReplayFromCheckpoint` arm of `restore_cross_turn_prefix` (lines 16151-16184) is a hard error:

```rust
PrefixReuseMode::ReplayFromCheckpoint { .. } => Err(InferenceError::PrefixCache(
    "restore_cross_turn_prefix: sparse GDN checkpoint replay is not implemented in v1".into(),
)),
```

Net effect: cache reuse only ever helps the one case where the new prompt is an exact append onto the single retained (latest) turn. Any conversation shape that shares a prefix with an *earlier* point than that — editing or regenerating an earlier turn, an agent branching to try an alternate continuation, or recovering after an interrupted turn — gets zero reuse and falls all the way back to `FullRefill` (full re-prefill), even though the planner is already written to handle it.

## Acceptance criteria

- [ ] Retain a small bounded ring of sparse GDN checkpoints per cross-turn slot (not just the single latest boundary), each tagged with its represented token length.
- [ ] Feed `plan_cross_turn_reuse` the slot's real checkpoint lengths instead of `&[]`.
- [ ] Implement the `ReplayFromCheckpoint` arm of `restore_cross_turn_prefix`: restore GDN state (`restore_gdn_states_checked`, line 16345) + KV to `checkpoint_len`, then prefill the divergent suffix forward from there using the existing arbitrary-`start_pos` suffix-prefill primitive (`forward_prefill_from`, line 16199).
- [ ] Bound checkpoint-ring memory and document the memory/turn-count tradeoff (a single GDN snapshot is ~19MiB per #516's measurement).
- [ ] Add a test alongside `cross_turn_cache_invalidated_by_gdn_mode_flip` (metal_qwen35.rs:24120) that verifies a replayed-from-checkpoint continuation produces logits byte-identical to a full re-prefill of the same tokens.

## Notes

Builds on the merged #516, which shipped `ExactAppend` + `FullRefill` only. Distinct from #462 (wiring the *existing* modes into `chat_metal.rs`/`lattice_serve.rs` — explicitly called out as an out-of-scope follow-up ("D6") in #516's own remediation notes, and still open for that reason) and from #486 (a public snapshot/branch/rollback API for agentic tool-call control flow — a different consumer layer per that issue's own "Distinction" section). Tracked as a sub-work item under the #481 parent ("Cross-turn / cross-request state caching"). Related prior art: Marconi's exact-match checkpoint admission/eviction model (arXiv:2411.19379).
